shadow/src/Nixfmt/Parser.hs

```diff
@@ -0,0 +1,317 @@
+{- © 2019 Serokell <hi@serokell.io>
+ - © 2019 Lars Jellema <lars.jellema@gmail.com>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+{-# LANGUAGE LambdaCase, OverloadedStrings #-}
+
+module Nixfmt.Parser where
+
+import Prelude hiding (String)
+
+import Control.Monad
+import qualified Control.Monad.Combinators.Expr as MPExpr
+import Data.Char
+import Data.Foldable (toList)
+import Data.Maybe
+import Data.Text as Text hiding (concat, concatMap, filter, init, last, map)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char (char)
+import qualified Text.Megaparsec.Char.Lexer as L
+
+import Nixfmt.Lexer
+import Nixfmt.Types
+import Nixfmt.Util
+
+-- HELPER FUNCTIONS
+
+ann :: (a -> b) -> Parser a -> Parser (Ann b)
+ann f p = try $ lexeme $ f <$> p
+
+-- | parses a token without parsing trivia after it
+rawSymbol :: Token -> Parser Token
+rawSymbol t = chunk (tokenText t) *> return t
+
+symbol :: Token -> Parser (Ann Token)
+symbol = lexeme . rawSymbol
+
+reservedNames :: [Text]
+reservedNames =
+    [ "let", "in"
+    , "if", "then", "else"
+    , "assert"
+    , "with"
+    , "rec"
+    , "inherit"
+    ]
+
+reserved :: Token -> Parser (Ann Token)
+reserved t = try $ lexeme $ rawSymbol t
+    <* lookAhead (satisfy (\x -> not $ identChar x || pathChar x))
+
+-- VALUES
+
+integer :: Parser (Ann Token)
+integer = ann Integer L.decimal
```

Comment:
Doesn't nix allow negative numbers?

Reply:
It does, but they are simply parsed as a negation expression wrapping a positive number.